apps/src/applab/RecordListener.js

```diff
@@ -0,0 +1,230 @@
+ * @returns {boolean} true if adding the listener succeeded, or false if
+ * a listener already existed for the specified table.
+ */
+RecordListener.prototype.addListener = function (tableName, callback) {
+  if (this.tableHandlers_[tableName]) {
+    return false;
+  }
+
+  this.tableHandlers_[tableName] = new TableHandler(tableName, callback);
+
+  return true;
+};
+
+/**
+ * Clears the timeouts and resets internal state.
+ */
+RecordListener.prototype.reset = function () {
+  for (var tableName in this.tableHandlers_) {
+    this.tableHandlers_[tableName].reset();
+  }
+  this.tableHandlers_ = {};
+};
+
+//////////////////////////////////////////////////
+// TableHandler
+//////////////////////////////////////////////////
+
+/**
+ * This class encapsulates all the logic and state required for listening to
+ * changes to a single table.
+ * @param {string} tableName
+ * @param {function} callback
+ * @constructor
+ */
+var TableHandler = function (tableName, callback) {
+  /**
+   * @private {string}
+   */
+  this.tableName_ = tableName;
+
+  /**
+   *  @private {function}
+   */
+  this.callback_ = callback;
+
+  /**
+   * @typedef {Object.<number, string>} IdToJsonMap Map from record id
+   * to a JSON-stringified copy of the corresponding record.
+   */
+
+  /**
+   * IdToJsonMap representing the contents of the table.
+   * We use this to efficiently test whether a record's contents has changed.
+   * @private {IdToJsonMap}
+   */
+  this.idToJsonMap_ = {};
+
+  /**
+   * Timeout id of the pending call to window.setTimeout.
+   * @private {number}
+   */
+  this.timeoutId_ = null;
+
+  this.scheduleNextFetch_();
+};
+
+/**
+ * @private
+ */
+TableHandler.prototype.scheduleNextFetch_ = function () {
+  this.timeoutId_ = setTimeout(function recordListener() {
+    this.fetchRecords_();
+  }.bind(this), RECORD_INTERVAL);
+};
+
+/**
+ * @private
+ */
+TableHandler.prototype.fetchRecords_ = function () {
+  var req = new XMLHttpRequest();
+  req.onreadystatechange = this.handleFetchRecords_.bind(this, req);
+  var url = '/v3/shared-tables/' + Applab.channelId + '/' + this.tableName_;
+  req.open('GET', url, true);
+  req.send();
+};
+
+/**
+ * @param {XMLHttpRequest} req
+ * @private
+ */
+TableHandler.prototype.handleFetchRecords_ = function (req) {
+  var done = XMLHttpRequest.DONE || 4;
+  if (req.readyState !== done) {
+    return;
+  }
+
+  if (req.status < 200 || req.status >= 300) {
+    // Ignore errors. In the future we may want to add monitoring or notify
+    // the user that this is happening.
+    return;
+  }
+
+  if (!this.idToJsonMap_) {
+    // The maps may have been reset while we were fetching the data.
+    return;
+  }
+
+  // Performance notes: on a table with 2000 records on localhost it takes ~1.5 seconds
+  // to fetch the data from dynamodb and 30ms to process it here. Processing time may be
+  // longer on slower machines. One quick and dirty performance improvement is to cache
```

Comment:
It would be ideal for the client to specify an If-Modified-Since header on its request and have the server return 304 if there have been no changes to the table.  This would avoid sending the whole table down the wire on every request which is a large expense even if you don't parse it.

When the table is returned it would include a datastamp to use for future If-Modified-Since requests so that we don't care about inaccurate client clocks.

Note that t